### ISSUE TITLE
AES encrypt-then-decrypt does not produce original file

### ISSUE THREAD

**Issue by idodeclare** [omniosorg/illumos-omnios#657]
Please see below chat from 31-Dec-2019 where jbk diagnosed a bug in OmniOS encrypt/decrypt and identified the illumos-gate commit that would fix it:

```
[18:50:31] <idodeclare> I'm not sure when this broke but `encrypt -a aes ...` followed by `decrypt -a aes ...` does not produce the original input file
[19:00:29] <LeftWing> idodeclare: It'd help to produce a short shell program that demonstrates the failure
[19:03:28] <jbk> also, kernel/distro versions
[19:04:19] <jbk> the encrypt/decrypt commands used CKM_AES_CBC_PAD, there were some bugs that were fixed this past year
[19:04:50] <jbk> (so good to know if you are seeing the issue with/without the fixes)
[19:10:46] <jbk> (having found and looked at the source for them)
[19:25:32] <idodeclare> LeftWing: OK https://gist.github.com/idodeclare/5b07f5143d9115e3ce4128938895a55a to encrypt/decrypt /etc/resolv.conf. Run ./demo_encrypt_decrypt arcfour to show a working algo and then ./demo_encrypt_decrypt aes
[19:25:44] <idodeclare> jbk: I'm on latest Omni OS r151032
[19:29:27] <jbk> it looks like it's missing this commit:
[19:29:30] <jbk> commit 8d91e49dd95381d46f9364f5de9e9027a11e1118
[19:29:30] <jbk> Author: Jason King <jason.king at joyent dot com>
[19:29:30] <jbk> Date:   Fri Jun 28 00:45:40 2019 +0000
[19:29:30] <jbk>     11825 PKCS#11 CKM_AES_CBC_PAD decryption can fail
[19:29:30] <jbk>     Reviewed by: Dan McDonald <danmcd at joyent dot com>
[19:29:32] <jbk>     Approved by: Gordon Ross <gordon.w.ross at gmail dot com>
[19:30:24] <jbk> I've got to run and do a few things, but I can try that a bit later on the latest smartos (which has that fix)
[19:31:21] <jbk> which would tell us if that does fix the issue (I suspect it will, but always nice to be able to confirm it)
[19:31:36] <idodeclare> jbk: thank you for looking at it
[19:32:17] <jbk> well i accidentialy broke it trying to add support for other stuff (sorry about that :( ), so only fitting I should try to fix it
[19:32:57] <jbk> unfortunately, the padding requirements and the PKCS#11 API requirements intersect in some unfortunate and annoying ways
[19:42:07] <idodeclare> Oh :) well much appreciated
[20:56:03] <jbk> idodeclare: looks like on the latest smartos it works ok..
[20:57:09] <jbk> i'm not sure what the omnios policy is on backporting fixes, though the person most likely to know is in the Uk and is likely further along on the NYE celebrations than here in the US :), so it may be a day or two before someone knows the answer
[20:58:59] <jbk> https://pastebin.com/5zz9UrR2
[21:02:46] <jbk> if you're in a pinch and need something sooner, I can probably build you a fixed libpkcs11_softtoken.so (that's where the bug is) that you could replace or loopback mount to make the encrypt/decrypt commands work w/ AES
[21:06:11] <idodeclare> jbk: oh that's encouraging. No I'm not in a pinch — but thank you for offering. Happy New Year!
[21:13:52] <jbk> ok.. if that changes feel free to let me know.. and sorry again for the breakage
[21:16:39] <idodeclare> no worries!
```

**Comment by hadfl** [comment]
there is a hot-fix available for r151032:

`pkg apply-hot-fix https://hf.omniosce.org/r151032/11825.p5p`

**Comment by idodeclare** [comment]
OK I see. Thank you.

BTW I had hot-fixed r151030 and was stymied when trying to upgrade to r151032. I don't exactly remember what I had to do to get unblocked, but I think it was needing to remove the hot fix publisher. May I recommend adding a section to [Upgrading to a New OmniOS Release](https://omniosce.org/upgrade.html) to address previous hot-fixes?

**Comment by hadfl** [comment]
hot-fix publishers are added temporarily; so after the hot-fix is applied, the publisher is removed again. if e.g. the user cancels apply-hot-fix in action it can be that there is a stray publisher in the ips image which needs to be removed manually.

having said that: this fix will be part of the release scheduled for beginning of this week

**Comment by idodeclare** [comment]
> hot-fix publishers are added temporarily; so after the hot-fix is applied, the publisher is removed again. if e.g. the user cancels apply-hot-fix in action it can be that there is a stray publisher in the ips image which needs to be removed manually.

Oh curious

**Comment by idodeclare** [comment]
I applied the hot-fix, but I did not perceive it fixing the problem. The sample program, [demo_encrypt_decrypt.sh](https://gist.github.com/idodeclare/5b07f5143d9115e3ce4128938895a55a), I wrote at LeftWing's request does not show it successfully encrypting/decrypting to the original file (/etc/resolv.conf) for the `aes` algorithm.

E.g. `./demo_encrypt_decrypt.sh des` works to round-trip the file, but `./demo_encrypt_decrypt.sh aes` shows gibberish.
